source/slang/glsl.meta.slang

```diff
@@ -2821,6 +2821,453 @@ public vec4 shadow2DProjLod(sampler2DShadow sampler, vec4 coord, float lod)
+        StringBuilder fullTypeNameBuilder;
+        fullTypeNameBuilder << targetType.prefix << "image" << targetShape.suffix;
+        auto fullTypeName = fullTypeNameBuilder.toString();
+
+        // We need a string to resolve for the imageSize return type
+        // since each shape has a different return which may not be
+        // a vec
+        StringBuilder imageSizeIntOrIVecBuilder;
+        if (targetShape.imageSizeIVecDim == 1)
+            imageSizeIntOrIVecBuilder << "int";
+        else
+            imageSizeIntOrIVecBuilder << "ivec" << targetShape.imageSizeIVecDim;
+        auto imageSizeIntOrIVec = imageSizeIntOrIVecBuilder.toString();
+
+        // Following OpenGL/Vulkan semantics, IMAGE_PARAMS string represents
+        // the changing list of parameters per gimageDim type
+        // https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.60.pdf
+        StringBuilder IMAGE_PARAMSBuilder;
+        IMAGE_PARAMSBuilder << fullTypeName << " image,";
+        int paramC = targetShape.imageCoordIVecIndexerDim;
+        if (paramC == 1)
+            IMAGE_PARAMSBuilder << "int";
+        else
+            IMAGE_PARAMSBuilder << "ivec" << paramC;
+        IMAGE_PARAMSBuilder << " P";
+        if (targetShape.isMS)
+        {
+            IMAGE_PARAMSBuilder << ", int sample";
+        }
+        auto IMAGE_PARAMS = IMAGE_PARAMSBuilder.toString();
+
+        // SPIR-V requires a 'sample' parameter for all texture
+        // functions. We need to handle a dynamic creation of
+        // 'sample' if 'sample' parameter is not in IMAGE_PARAMS
+        StringBuilder SPV_PREFIX_IMAGE_PARAMSBuilder;
+        StringBuilder SPV_SUFFIX_IMAGE_PARAMSBuilder;
+        StringBuilder SPV_DEFAULT_SAMPLE_VAR_IF_MISSINGBuilder;
+
+        SPV_PREFIX_IMAGE_PARAMSBuilder << "$image $P";
+        SPV_SUFFIX_IMAGE_PARAMSBuilder;
+        if (targetShape.isMS)
+        {
+            SPV_SUFFIX_IMAGE_PARAMSBuilder << " Sample $sample";
+        }
+        else
+        {
+            SPV_DEFAULT_SAMPLE_VAR_IF_MISSINGBuilder << "let sample = 0;";
+        }
+
+        auto SPV_PREFIX_IMAGE_PARAMS = SPV_PREFIX_IMAGE_PARAMSBuilder.toString();
+        auto SPV_SUFFIX_IMAGE_PARAMS = SPV_SUFFIX_IMAGE_PARAMSBuilder.toString();
+        auto SPV_DEFAULT_SAMPLE_VAR_IF_MISSING = SPV_DEFAULT_SAMPLE_VAR_IF_MISSINGBuilder.toString();
+    }}}}
+
+        __generic<let format:int=0>
+        public typealias $(fullTypeName) = __TextureImpl<
+            $(targetType.type),
+            $(targetShape.shape),
+            $(targetShape.isArray), // isArray
+            $(targetShape.isMS), // isMS
+            0, // sampleCount
+            1, // access
+            0, // isShadow
+            0, // isCombined
+            format
+        >;
+
+${{{{
+        if(targetShape.isRect)
+            continue;
+}}}}
+
+        [require(spirv)]
+        [require(glsl)]
+        [ForceInline] public $(imageSizeIntOrIVec) imageSize(readonly writeonly $(fullTypeName) image)
```

Comment:
`readonly writeonly` doesn't make much sense?

Reply:
its glsl syntax for "no side effect, no reading, glsl is doing something with the variable, not with it's data"

Reply:
Ah, that reminds me we should also decorate these functions with `[__readNone]` to indicate the function has no side effects and can be removed if its result is not used.

Reply:
The OpenGL spec says this,
```
A variable could be qualified as both readonly and writeonly, disallowing both read and write.
Such variables can still be used with some queries, for example imageSize() and .length()
```
Yong and I asked the same question.
And probably whoever will look at this code will have the same question.
Can you leave a comment about this to be more efficient for the readers?